stratosphere/loader/source/ldr_process_creation.cpp

```diff
@@ -14,248 +14,698 @@
+
+            /* Set out. */
+            aslr_start += aslr_slide;
+            for (size_t i = 0; i < Nso_Count; i++) {
+                if (has_nso[i]) {
+                    out->nso_address[i] += aslr_start;
+                }
+            }
+            if (out->args_address) {
+                out->args_address += aslr_start;
+            }
+
+            out_cpi->code_address = aslr_start;
+            out_cpi->code_num_pages = total_size >> 12;
+
+            return ResultSuccess;
         }
-    };
 
-    /* Load the process's NPDM. */
-    R_TRY(NpdmUtils::LoadNpdmFromCache(target_process->tid_sid.title_id, &npdm_info));
+        Result CreateProcessImpl(ProcessInfo *out, const Meta *meta, const NsoHeader *nso_headers, const bool *has_nso, const args::ArgumentInfo *arg_info, u32 flags, Handle reslimit_h) {
+            /* Get CreateProcessInfo. */
+            CreateProcessInfo cpi;
+            R_TRY(GetCreateProcessInfo(&cpi, meta, flags, reslimit_h));
 
-    /* Validate version. */
-    R_TRY(ValidateProcessVersion(target_process->tid_sid.title_id, npdm_info.header->version));
+            /* Decide on an NSO layout. */
+            R_TRY(DecideAddressSpaceLayout(out, &cpi, nso_headers, has_nso, arg_info));
 
-    /* Validate the title we're loading is what we expect. */
-    if (npdm_info.aci0->title_id < npdm_info.acid->title_id_range_min || npdm_info.aci0->title_id > npdm_info.acid->title_id_range_max) {
-        return ResultLoaderInvalidProgramId;
-    }
+            /* Actually create process. const_cast necessary because libnx doesn't declare svcCreateProcess with const u32*. */
+            return svcCreateProcess(out->process_handle.GetPointer(), &cpi, const_cast<u32 *>(reinterpret_cast<const u32 *>(meta->aci_kac)), meta->aci->kac_size / sizeof(u32));
+        }
 
-    /* Validate that the ACI0 Kernel Capabilities are valid and restricted by the ACID Kernel Capabilities. */
-    const u32 *acid_caps = reinterpret_cast<u32 *>(npdm_info.acid_kac);
-    const u32 *aci0_caps = reinterpret_cast<u32 *>(npdm_info.aci0_kac);
-    const size_t num_acid_caps = npdm_info.acid->kac_size / sizeof(*acid_caps);
-    const size_t num_aci0_caps = npdm_info.aci0->kac_size / sizeof(*aci0_caps);
-    R_TRY(NpdmUtils::ValidateCapabilities(acid_caps, num_acid_caps, aci0_caps, num_aci0_caps));
+        Result LoadNsoSegment(FILE *f, const NsoHeader::SegmentInfo *segment, size_t file_size, const u8 *file_hash, bool is_compressed, bool check_hash, uintptr_t map_base, uintptr_t map_end) {
+            /* Select read size based on compression. */
+            if (!is_compressed) {
+                file_size = segment->size;
+            }
 
-    /* Read in all NSO headers, see what NSOs are present. */
-    R_TRY(NsoUtils::LoadNsoHeaders(npdm_info.aci0->title_id));
+            /* Validate size. */
+            if (file_size > segment->size || file_size > std::numeric_limits<int>::max() || segment->size > std::numeric_limits<int>::max()) {
```

Comment:
Regular primitive `int` type here, what's the reason for that?

Reply:
It implements this check from real loader:

![image](https://user-images.githubusercontent.com/8676005/60309291-e5be9400-9901-11e9-9bff-17521fff407b.png)

Reply:
I meant why int versus s32. Is it related to usage of stdio/etc?